pkg/kms/remotekms/remotekms.go

```diff
@@ -0,0 +1,266 @@
+)
+
+const (
+	// Namespace of the remoteKMS local config storage.
+	Namespace = "remotekmsdb"
+
+	// KeystoreURLField representing the user's keystore URL field in the remoteKMS local storage.
+	KeystoreURLField = "keystoreURL"
+
+	// ControllerField representing the user's controller field in the remoteKMS local storage.
+	ControllerField = "controller"
+
+	// SecretField representing the user's secret field in the remoteKMS local storage.
+	SecretField = "secret"
+
+	createKeystoreReq = `{
+	  "controller": "%s"
+	  "passphrase": "%s"
+	}`
+
+	createKeyReq = `{
+	  "keyType": "%s",
+	  "passphrase": "%s"
+	}`
+
+	exportKeyReq = `{
+	  "keyID": "%s",
+	  "passphrase": "%s"
+	}`
+
+	createKeystoreEndpoint = "{serverEndpoint}/kms/keystores"
+
+	// ContentType is remoteKMS http content-type.
+	ContentType = "application/json"
+
+	// LocationHeader is remoteKMS http header set by the key server (usually to identify a keystore or key url).
+	LocationHeader = "Location"
+
+	// KeyBytesHeader is remoteKMS http header set by the key server (to contain an base64URL encoded public key).
+	KeyBytesHeader = "KeyBytes"
+)
+
+var logger = log.New("aries-framework/remotekms")
+
+// RemoteKMS implementation of kms.KeyManager api.
+type RemoteKMS struct {
+	httpClient  *http.Client
+	secret      string
+	configStore storage.Store
+	keystoreURL string
+}
+
+// New creates a new remoteKMS instance using http client connecting to a server at a url saved in a local remoteKMS
+// configuration configStore called as in Namespace const.
+func New(provider storage.Provider, client *http.Client, secret string) (*RemoteKMS, error) {
+	configStore, err := provider.OpenStore(Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	keystoreURL, err := configStore.Get(KeystoreURLField)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch keystore url from remoteKMS storage: %s", err)
+	}
+
+	return &RemoteKMS{
+		httpClient:  client,
+		secret:      secret,
+		configStore: configStore,
+		keystoreURL: string(keystoreURL),
+	}, nil
+}
+
+// CreateKeyStore calls the key server's create keystore REST function and saves the resulting keystoreURL value in a
+// local remoteKMS configStore.
+// Arguments of this function are described below:
+//   - configStore is where the keystoreURL, controller and secret will be stored
+//   - httpClient used to POST the request
+//   - keyserverURL representing the key server url
+//   - controller is the identifier of the keystore owner (usually a DID key ID)
+//   - secret represents the configStore secret for locking/unlocking the keystore
+// Returns:
+//  - keystore URL (if successful)
+//  - error (if error encountered)
+func CreateKeyStore(store storage.Store, httpClient *http.Client, keyserverURL, controller,
+	secret string) (string, error) {
+	destination := strings.ReplaceAll(createKeystoreEndpoint, "{serverEndpoint}", keyserverURL)
+	req := []byte(fmt.Sprintf(createKeystoreReq, controller, secret))
```

Comment:
Can we use Go struct to define the req/response model and use json marshal ? It would be hard to maintain with string formatting to create the json requests.

Reply:
sure.. the same goes for the key server @aholovko ;-)

I will convert createKeyStoreReq to be json objects and pass them directly

Reply:
done